### PR TITLE
Add Formatter supertype

### PR DESCRIPTION
**Pull request by sp94**
Add a generic Formatter type for QuantumarticleFormatter and RevtexFormatter to inherit from. This reduces the amount of repeated code and makes it easier to add formatters for new journals.

### CODE REVIEW

**Comment by johannesjmeyer** [comment]
Very cool! I'll review it ASAP.

**Comment by johannesjmeyer** [comment]
First of all, thanks a lot for the PR! This will help a lot to streamline the code.

Second, an explanation: the document classes `revtex` and `quantumarticle` are very similar, basically because `quantumarticle` was designed to be 100% compatible to `revtex` in terms of its arguments. This is also why they share so much common code.
But a word of caution is advised: other document classes will not have the same kinds of arguments, which means that they will not integrate well with the current `Formatter` class. As `rsmf` should not stay limited to `revtex` and `quantumarticle` in the future, a more abstract base is needed. 

So therefore my first suggestion: The current `Formatter` class will be renamed to something along the lines of `RevtexLikeFormatter` (didn't come up with a better name yet, but it actually covers quite well what `quantumarticle` is). Then, another class `Formatter` that basically only has the interface methods and sets `pgf` as the backend should be introduced.

Third, parsing. I saw that you simplified the parsing routine in `setup.py`. But the way it is now it will only work for `revtex` and `quantumarticle` and is not extensible to other document classes which have different arguments. I built it the way it was because I had this extensibility in mind, along with more complicated future use cases where the whole preamble could be parsed, e.g. to look for additional font-changing packages. I also like the idea of providing an interface to `setup` whose sole purpose is to take the preamble and give me the appropriate formatter, which is done by the parsing functions. 

I suggest we roll back to the original logic and unify the two parsers using your code and a functor like
```python
class RevtexLikeParser:
    def __init__(self, documentclass_identifiers):
        self.documentclass_identifiers = documentclass_identifiers

    def __call__(self, preamble):
        # the old boilerplate here, testing for all given identifiers

revtex_parser = RevtexLikeParser(["{revtex4-1}", "{revtex4-2}"])
quantumarticle_parser = RevtexLikeParser(["{quantumarticle}"])
```
I also just realized that `revtex4-1` and `revtex4-2` are both valid, hence the array.
___
What do you think @sp94 ?

File: rsmf/formatters.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+    def __eq__(self, other):
+        return (
+            self.fontsize == other.fontsize
+            and self.columns == other.columns
+            and self.paper == other.paper
+        )
+
+
+def parse(preamble):
```
Comment (johannesjmeyer) [inline rsmf/formatters.py]: I guess this is meant to be removed?

File: rsmf/quantumarticle.py [hunk 2]
```diff
@@ -5,34 +5,29 @@
...
-class QuantumColors:
-    """Special colors defined in the quantumarticle document class."""
+    _wide_widths = {
+        "onecolumn": {"a4paper": 5.93, "letterpaper": 6.16},
+        "twocolumn": {"a4paper": 6.72, "letterpaper": 6.95},
+    }
 
-    quantumviolet = "#53257F"
-    quantumgray = "#555555"
+    _colors = {
```
Comment (johannesjmeyer) [inline rsmf/quantumarticle.py]: Definitely makes more sense as a dict!

File: rsmf/formatters.py (hunk 1 above)
Comment (sp94) [reply rsmf/formatters.py]: Oops, yes!

**Comment by johannesjmeyer** [comment]
I also have to check another thing: I just consulted the documentation of revtex and you can use it to submit to a whole host of non-APS journals, too, altering the style via separate commands. If these change the formatting and column widths we can use the new class structure to add additional subclasses for 'RevtexLikeFormatter'

**Comment by sp94** [comment]
OK, I think I see your purpose with the original design!

> So therefore my first suggestion: The current Formatter class will be renamed to something along the lines of RevtexLikeFormatter (didn't come up with a better name yet, but it actually covers quite well what quantumarticle is). Then, another class Formatter that basically only has the interface methods and sets pgf as the backend should be introduced.

OK, so the most general `Formatter` should only have `width`, `wide_width` and `fontsizes` as its properties?

> Third, parsing. I saw that you simplified the parsing routine in setup.py. But the way it is now it will only work for revtex and quantumarticle and is not extensible to other document classes which have different arguments. I built it the way it was because I had this extensibility in mind, along with more complicated future use cases where the whole preamble could be parsed, e.g. to look for additional font-changing packages. I also like the idea of providing an interface to setup whose sole purpose is to take the preamble and give me the appropriate formatter, which is done by the parsing functions.

I see, so really parsing the preamble differently for different documentclasses is the hardest part.

Can I suggest that rather than putting `documentclass_identifiers` in the parser, we put it in the `Formatter`? Something like

```python
class Formatter:

    @classmethod
    def from_preamble(cls, preamble):
        for id in self.documentclass_identifiers:
            if id in preamble:
                formatter_kwargs = self.parser.parse(preamble)
                return cls(**kwargs)
```

Then if you want to create a new formatter using an existing parser, this is all you have to do:
```python
class QuantumarticleFormatter(Formatter):
    parser = RevtexParser

    def __init__(self, columns="twocolumn", paper="a4paper", fontsize=10, **kwargs):
        # sets width, wide_width, fontsizes
        # maybe overrides some rcParams

# and then inside your for loop of setup.py:
formatter = QuantumarticleFormatter.from_preamble(preamble)
```

**Comment by johannesjmeyer** [comment]
> OK, so the most general `Formatter` should only have `width`, `wide_width` and `fontsizes` as its properties?

Yes, along with all the code that is sufficiently general (i.e. figure generation) and some basic rcParams.

> I see, so really parsing the preamble differently for different documentclasses is the hardest part.
> 
> Can I suggest that rather than putting `documentclass_identifiers` in the parser, we put it in the `Formatter`? Something like
> 
> ```python
> class Formatter:
> 
>     @classmethod
>     def from_preamble(cls, preamble):
>         for id in self.documentclass_identifiers:
>             if id in preamble:
>                 formatter_kwargs = self.parser.parse(preamble)
>                 return cls(**kwargs)
> ```

I would rather keep the parser separate. There is no functional difference, it's just a design decision. 

I'd modify your PR as soon as I can spare the time and then you can have another look at it, how does that sound?

**Comment by sp94** [comment]
Alright, sounds good!

**Comment by johannesjmeyer** [comment]
@sp94 you can have another look and review? I think I made all the changes. I also restructured the fonts handling a bit (while I don't think reasonable change the standard LaTeX font hierarchy, the new approach is more flexible towards that end by just having a Fontsizes class that has all the named fontsizes as attributes).

**Comment by johannesjmeyer** [comment]
I also added some TODOs I wanted to check in a separate PR before the next release. 

By the way, I've seen that you do research into photonic crystals. In what journals besides APS does this research get published? It would be cool to expand the available publication venues with things people care about.

**Comment by sp94** [comment]
I just submitted to APS (wish I'd seen this package before that...!) but I'm currently looking at the package for use with the Classicthesis template

Edit: So for my personal use case I'm only looking for the sizing of the figures and text, because as its thesis and not going to a journal the guidelines aren't so strict, and I'm pretty happy with the matplotlib defaults for most things

**Comment by johannesjmeyer** [comment]
That's actually quite cool, because that will also be very useful to many I guess. After this PR is through we should implement that!

File: rsmf/revtexlike.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        """Fontsizes as specified by the underlying document."""
+        return default_fontsizes[self.fontsize]
+
+    def set_default_fontsizes(self):
+        """Adjust the fontsizes in rcParams to the default values matching the surrounding document."""
+        plt.rcParams["axes.labelsize"] = self.fontsizes.small
+        plt.rcParams["axes.titlesize"] = self.fontsizes.large
+        plt.rcParams["xtick.labelsize"] = self.fontsizes.footnotesize
+        plt.rcParams["ytick.labelsize"] = self.fontsizes.footnotesize
+        plt.rcParams["font.size"] = self.fontsizes.small
```
Comment (sp94) [inline rsmf/revtexlike.py]: Is this already covered by the base class?

**Comment by sp94** [comment]
I removed what I think was a duplicated method in `RevtexLikeFormatter` but otherwise looks good to go!

I have some other questions about whether eg Formatter should be usable by itself, but I'll open a new issue for that

**Pull request merged**